js/src/offcanvas.js

```diff
@@ -0,0 +1,194 @@
+const VERSION = '4.3.1'
+const DATA_KEY = 'bs.offcanvas'
+const EVENT_KEY = `.${DATA_KEY}`
+const DATA_API_KEY = '.data-api'
+
+const Selector = {
+  DATA_DISMISS: '[data-dismiss="offcanvas"]',
+  DATA_TOGGLE: '[data-toggle="offcanvas"]'
+}
+
+const Event = {
+  SHOW: `show${EVENT_KEY}`,
+  SHOWN: `shown${EVENT_KEY}`,
+  HIDE: `hide${EVENT_KEY}`,
+  HIDDEN: `hidden${EVENT_KEY}`,
+  CLICK_DATA_API: `click${EVENT_KEY}${DATA_API_KEY}`,
+  CLICK_DISMISS: `click.dismiss${EVENT_KEY}`
+}
+
+const ClassName = {
+  SHOW: 'show'
+}
+
+/**
+ * ------------------------------------------------------------------------
+ * Class Definition
+ * ------------------------------------------------------------------------
+ */
+
+class Offcanvas {
+  constructor(element) {
+    this._element = element
+    this._isShown = element.classList.contains(ClassName.SHOW)
+    if (this._element) {
+      Data.setData(element, DATA_KEY, this)
+    }
+
+    EventHandler.on(this._element, Event.CLICK_DISMISS, Selector.DATA_DISMISS, event => this.hide(event))
+    EventHandler.on(document, Event.CLICK_DATA_API, event => {
+      const target = Offcanvas._getParentFromElement(event.target)
+      if (!this._element.contains(event.target) && target !== this._element) {
+        this.hide(event)
+      }
+    })
+  }
+
+  // Getters
+
+  static get VERSION() {
+    return VERSION
+  }
+
+  // Public
+
+  toggle(relatedTarget) {
+    return this._isShown ? this.hide() : this.show(relatedTarget)
+  }
+
+  show(relatedTarget) {
+    if (this._isShown) {
+      return
+    }
+
+    const showEvent = EventHandler.trigger(this._element, Event.SHOW, {
+      relatedTarget
+    })
+
+    if (this._isShown || showEvent.defaultPrevented) {
+      return
+    }
+
+    this._isShown = true
+
+    this._element.classList.add(ClassName.SHOW)
+    this._element.removeAttribute('aria-hidden')
+    this._element.setAttribute('aria-offcanvas', true)
+    EventHandler.trigger(this._element, Event.SHOWN, {
+      relatedTarget
+    })
+  }
+
+  hide(event) {
+    if (event) {
+      event.preventDefault()
+    }
+
+    if (!this._isShown) {
+      return
+    }
+
+    const hideEvent = EventHandler.trigger(this._element, Event.HIDE)
+
+    if (!this._isShown || hideEvent.defaultPrevented) {
+      return
+    }
+
+    this._isShown = false
+
+    this._element.classList.remove(ClassName.SHOW)
+    this._element.setAttribute('aria-hidden', true)
+    this._element.removeAttribute('aria-offcanvas')
+
+    EventHandler.trigger(this._element, Event.HIDDEN)
+  }
+
+  // Static
+
+  static _getParentFromElement(element) {
+    let parent
+    const selector = getSelectorFromElement(element)
+
+    if (selector) {
+      parent = SelectorEngine.findOne(selector)
+    }
+
+    return parent || element.parentNode
+  }
+
+  static _jQueryInterface(config) {
+    return this.each(function () {
+      let data = Data.getData(this, DATA_KEY)
+
+      if (!data) {
+        data = new Offcanvas(this)
+      }
+
```

Comment:
it lacks of the check for undefined methods, it's better to allow jQuery folks to call other methods too
```js
    if (typeof config === 'string') {
      if (typeof data[config] === 'undefined') {
        throw new TypeError(`No method named "${config}"`)
      }

      data[config]()
    }

Reply:
I think it's ok now